sql: Align timestamp math with PostgreSQL

bobvawter opened the pull request:
Our implementation of time+duration math delegates to golang's
time.Time.AddDate(), which has different rounding semantics when compared to
PostgreSQL when normalizing invalid month-of-day values.

The most likely input case is that the duration will have no month component
and we can fall back to go's implementation.  Another fast-path of checking the
day-of-month allows us to skip the special-case logic for start dates that
aren't the 29th, 30th, or 31st of a month.

```
name                                      old time/op    new time/op    delta
Add/fast-path-by-no-months-in-duration-8    63.2ns ± 4%    66.2ns ± 1%    +4.75%  (p=0.016 n=5+5)
Add/fast-path-by-day-number-8               70.5ns ± 5%    91.1ns ± 1%   +29.20%  (p=0.008 n=5+5)
Add/no-adjustment-8                         70.6ns ± 3%   206.8ns ± 1%  +192.83%  (p=0.008 n=5+5)
Add/with-adjustment-8                       68.1ns ± 3%   228.4ns ± 1%  +235.59%  (p=0.008 n=5+5)

(No difference in allocations)
```

Resolves #30976

Release note (sql change): The normalized results of certain timestamp +
duration operations involving year or month durations have been adjusted to
agree with the values returned by PostgreSQL.  In cases such as
'2018-01-31'::TIMESTAMP + '1 month', where an intermediate result of February
31st needs to be normalized, previous versions of CockroachDB would "round up"
to March 1.  Instead, we would now "round down" to February 28th to agree with
the values returned by PostgreSQL.  This change also affects the results of
`generate_sequence()` when used with timestamps.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/31146)
<!-- Reviewable:end -->

knz commented:
cc @dt

knz commented on pkg/util/duration/duration.go:
```diff
@@ -265,10 +265,61 @@ func Decode(sortNanos int64, months int64, days int64) (Duration, error) {
+// the behavior we see in PostgreSQL when starting from days
+// numbered 29, 30, or 31 and advancing yearly or monthly.  Instead
+// of normalizing February 31st into the next month, we need to
+// "round down" to the end of the month.
```
You have peeked into pg's implementation to determine what to do here. Please reference the particular pg version you looked at, and the filename inside the source tree you looked at, and the C function name(s) (or tests), in the comment.

knz reviewed (approved):
LGTM but would welcome additional reviewer's opinions.

bobvawter commented:
Does the timestamp-math change need to be hidden behind a setting or cluster version?  I’m wondering what happens if you run `select some_timestamp + '1 month' from table`.  Is more than one node going to be evaluating that addition?

knz commented:
Yes more than 1 node is going to be evaluating this. This is an "interesting" question.

knz commented:
@jordanlewis do you want to provide guidance about who to talk to to make this behavior properly conditional?

jordanlewis commented:
Hmm. Bumping the DistSQL minimum version will prevent problems with DistSQL, but it won't prevent the fact that, during a rolling upgrade, two nodes might give two different answers.

I believe the right thing to do is gate the new behavior on a cluster version. I don't remember how this is done.

knz commented:
@bobvawter here is an example of cluster version gate

```
pkg/sql/sqlbase/structured.go:   
       if !st.Version.IsMinSupported(cluster.VersionBitArrayColumns) {
```

knz commented:
I would recommend a (hidden) session setting  in SessionData
which is initialized when the session starts

(SEssionData variables are transferred across distsql flows)

danhhz commented:
I haven't been in the duration stuff (or even thought about it) in ~2 years, so going to remove myself from the review, if that's okay.

bobvawter commented:
Is the SessionData or cluster version available from a static context, or is this going to require a ton of plumbing to get the necessary version down into `Add()`?

bobvawter commented:
It looks like there's an `EvalContext` available at many/most of the call-chains that eventually call into `Add()`.  I'll see what getting all of that plumbing set up looks like.

mjibson commented:
I'm not convinced we need cluster or session settings to handle this. Don't we already change implementation of functions regularly without gating them behind any kind of setting?

knz commented:
@mjibson that's a valid argument in general. -- if we could argue the previous behavior was unequivocally a bug , then you'd be right we don't need a flag.

However it's not so clear "on its face" that "Jan 31 + 1 month" should be equal to Feb 29 or March 1 (there are other places in postgres, and in cockroachdb, where 1 month = 30 days no matter what). There is an argument to be made in favor of either way, so I can't confidently say that the current behavior would be perceived as a bug by every user looking at it carefully.

bobvawter commented:
PTAL: I've updated this patch with a session variable and cluster setting to allow users to disable the new date math in case their app depends on the golang math.  I'm also using this as an opportunity to delay switching to the new logic until a cluster is fully upgraded.

bobvawter reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31146)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/util/duration/duration.go, line 275 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31146#-LOOsz2KW66lkV3HZHHT-r1-275:-LOQT3DL1gI8dyjaZbOH:bth8ysh) ([raw file](https://github.com/cockroachdb/cockroach/blob/083749621d3b6e3f672b84d16754780e073bb9fd/pkg/util/duration/duration.go#L275)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

You have peeked into pg's implementation to determine what to do here. Please reference the particular pg version you looked at, and the filename inside the source tree you looked at, and the C function name(s) (or tests), in the comment.
</blockquote></details>

Added this info to the doc on AdditionModeCompatible.


<!-- Sent from Reviewable.io -->

mjibson commented:
Here's a more thorough argument about why we should not have either a cluster version check or a session setting and instead just fix it and move on.

Let's begin by looking at the [recent changes in builtins](https://github.com/cockroachdb/cockroach/commits/master/pkg/sql/sem/builtins) that changed the results of a SQL function call (i.e., not fixes for panics):

- [sql: properly ignore invalid chars in encoding names](https://github.com/cockroachdb/cockroach/commit/c31e83ac72f2c8b9f8a6fae281071fb1e3306299#diff-eab6d12f93a40b49175acbbe6ff3a354)
- [sql: hashes with NULL input should return NULL](https://github.com/cockroachdb/cockroach/commit/1383db4229dbd3c6f82df6ac99f634b78d6a2a71#diff-eab6d12f93a40b49175acbbe6ff3a354)
- [sql: Fix the handling of NULL arguments in string_agg](https://github.com/cockroachdb/cockroach/commit/3546d64f342dd6846ed19b9c3db7958079cc426d#diff-081d6f4aea463b30b8f2909873d60426)
- [sql: correctly handle NULL arrays in generator funcs](https://github.com/cockroachdb/cockroach/commit/f0cd76cfca1ddd8237c2897789d9084e24dfa1b1#diff-081d6f4aea463b30b8f2909873d60426)

And some for [eval.go](https://github.com/cockroachdb/cockroach/commits/master/pkg/sql/sem/tree/eval.go):

- [sql: fix evaluation of <tuple> IS NULL](https://github.com/cockroachdb/cockroach/commit/222481882d8dd7690f03af56b57c463f12c7e6be#diff-12ba3fc0074d5c31d8c1518f691184ac)
- [sql: null array element in cast is null](https://github.com/cockroachdb/cockroach/commit/554a34c1d287a20ba5a42caec3ee0b86e96c76e2#diff-12ba3fc0074d5c31d8c1518f691184ac)

I think every one of the above changes would behave differently during a cluster upgrade, just like this one. I believe none of them had session setting or cluster version checks, we just called it a fixed bug and moved on with life. If someone was depending on the old behavior, too bad.

Another argument, a sort of proof by contradiction: I really *really* don't want to have a policy that we maintain with session settings any old behavior that doesn't match postgres in addition to the new behavior that does match it. For example, if we are changing the dates as above to make it match postgres and introduce various settings to allow users to choose which implementation they want, would we also not do the same for, say, integer division? We made the choice years ago to have `int / int` return a `decimal` instead of an `int` like postgres does. I think if we apply the same thought process about the date math above to integer division, we would have to change it to match postgres and add a setting to allow for the previous behavior. I think it would be a bad choice to do this (have a setting) for int division, so I also think it's a bad choice for date math.

(I am in full support of just changing int division to match postgres if our policy is to just match postgres during edge cases. I'm equally happy to leave it as is.)

knz commented:
Matt your second argument about the maintenance burden is the strongest argument I think. I find i very compelling.

For the sake of completeness I'd like to set the record clear about the examples you've chosen:

- sql: properly ignore invalid chars in encoding names

  This is removing errors where previously there was an error. So it's a behavior extension not a backward-incompatible change.

- sql: hashes with NULL input should return NULL

  Anything that changes/improves the behavior with NULL values should be considered with the same criteria as bug fixes in error codes / error messages. NULL is not/rarely used in SQL for anything else than ignoring errors.

- sql: Fix the handling of NULL arguments in string_agg

  ditto

- sql: correctly handle NULL arrays in generator funcs

  ditto

- sql: fix evaluation of IS NULL

  ditto. Moreover, NULL has the same truthfulness as false in conditionals so queries using the "tuple is null" construct do not change semantics as a result of this change.

- sql: null array element in cast is null

  ditto

If you want to take examples that support your point I would recommend using those instead:

- sql: ceil no longer returns -0 for decimals

- sql: fix implementation of encode and decode

But as you can yourself confirm in the git log, it's actually pretty hard to find cases where we've made these type of value computation changes, and in all cases I've found so far the changes are on corner cases.

The changes proposed by Bob here are not on corner cases -- many "common values" for the arguments of the operators are impacted.

This is why the first part of your argument is not too convincing (for me). And also why I tend to concur with Ben that this change requires slightly more attention than our usual semantic fixes.

That said your argument on maintainability is pretty compelling, and that's also why we chose to ignore cross-version compatibility in the the 0/-0 debacle and the encode/decode story.

So perhaps we can do like Bob suggests, merge this without version check for now and revisit if a customer/user complains.

Finally, @bobvawter in any case please do include a "backward-incompatible change" release note in the commit message.

bobvawter commented:
Having thought about this overnight, I think we'd be creating one hell of a Heisenbug for developers to track down if we don't at least solve for the mixed-version cluster case since our old math and new math can differ by multiple days in some cases.

In the absence of data that shows that time+duration is rare, or that using month values is rare, I think we serve our users better by taking a more conservative approach in this case.

knz commented:
Bob I think I am OK with you taking that approach, matt's arguments notwithstanding, for two reason:

- I'll defer to your judgement about seriousness
- it's also a good exercise for us to understand how such a version conditional needs to be implemented in practice. You could consider this issue as an opportunity to teach something to the rest of the team we'll need to use in other serious cases.

bobvawter commented:
PTAL: Here's hoping this is the last rev.  After a discussion with Ben, this patch drops the user-configurable knobs, but is still version-aware to avoid inconsistencies within mixed-mode clusters.  The version check is performed once and memoized in the `SessionData`.

mjibson commented on pkg/settings/cluster/cockroach_versions.go:
```diff
@@ -283,6 +285,12 @@ var versionsSingleton = keyedVersions([]keyedVersion{
 		Key:     VersionLoadBasedRebalancing,
 		Version: roachpb.Version{Major: 2, Minor: 0, Unstable: 14},
 	},
+	// Version2_1 should go here.
```
I'm not 100% sure how this version stuff works. Do we need to wait until a 2.1.0 is made before making a 2.1.1?

bdarnell replied:
Yeah, things get a little weird if we do this out of order. Today's the day to make the 2.1 cluster version anyway. 

But I don't think it's a great idea to tie a change like this to the cluster version - the change happens at exactly the point where the upgrade becomes irreversible. I think it might be better to just make the change as soon as the new binary is used while rolling back is still possible.

mjibson commented on pkg/sql/sessiondata/session_data.go:
```diff
@@ -76,6 +77,8 @@ type SessionData struct {
 	// DataConversion gives access to the data conversion configuration.
 	DataConversion DataConversionConfig
+	// DurationAdditionMode enables math compatibility options to be enabled.
+	DurationAdditionMode duration.AdditionMode
```
Put a TODO here to remove this in the 2.2 release cycle, since at that point we know it's safe to remove the old behavior.

dt commented:
honestly you can still have a nasty heisenbug with drivers that connection pool since every statement you run might get a different session -- I'm increasingly in the "just rip the band-aid off" school.

bobvawter reviewed (commented):
@dt, true, but at least within any given execution it will be consistent.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31146#-:-LOUKhWM9Ej4KB6BZBim:b4qf1m0)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/settings/cluster/cockroach_versions.go, line 288 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/31146#-LOQJHfS9MbUUBOz0398-r2-288:-LOUKmOyEHFl8MJytTzc:bvzsw6e) ([raw file](https://github.com/cockroachdb/cockroach/blob/0017b1c26904f8168b7581528f59264edfa9e881/pkg/settings/cluster/cockroach_versions.go#L288)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Yeah, things get a little weird if we do this out of order. Today's the day to make the 2.1 cluster version anyway. 

But I don't think it's a great idea to tie a change like this to the cluster version - the change happens at exactly the point where the upgrade becomes irreversible. I think it might be better to just make the change as soon as the new binary is used while rolling back is still possible.
</blockquote></details>

I thought we wanted all of the nodes to agree that they can use the new date math before any of them start to do so to avoid inconsistency within any given session, but that we're ok with just changing the math behavior out from under the user.  Is there some other means to inform the decision about which mode to use in a SessionData?
___
*[pkg/sql/sessiondata/session_data.go, line 81 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/31146#-LOQJHfT0UKyfUT8KV8g-r2-81:-LOULHn_7pdYkLYLUE1W:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0017b1c26904f8168b7581528f59264edfa9e881/pkg/sql/sessiondata/session_data.go#L81)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Put a TODO here to remove this in the 2.2 release cycle, since at that point we know it's safe to remove the old behavior.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

bdarnell reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31146)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/settings/cluster/cockroach_versions.go, line 288 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/31146#-LOQJHfS9MbUUBOz0398-r2-288:-LOURibdCCLTGr1gpcNM:bg8nb0y) ([raw file](https://github.com/cockroachdb/cockroach/blob/0017b1c26904f8168b7581528f59264edfa9e881/pkg/settings/cluster/cockroach_versions.go#L288)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

I thought we wanted all of the nodes to agree that they can use the new date math before any of them start to do so to avoid inconsistency within any given session, but that we're ok with just changing the math behavior out from under the user.  Is there some other means to inform the decision about which mode to use in a SessionData?
</blockquote></details>

I think a distsql version is a good idea, which will prevent inconsistency within a session. 


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
I'm rather comfortable with the end result here. Just a minor nit about the generator interface and this will be good to co from my perspective.

> Reviewed 1 of 4 files at r1, 21 of 21 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31146#-:-LOUTWtA4SOQ0OmDGnEP:b-bzt5xy)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/sem/builtins/generator_builtins.go, line 260 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/31146#-LOUSzKB8MWLq2r_PSxU:-LOUSzKB8MWLq2r_PSxV:b-wjlqd7) ([raw file](https://github.com/cockroachdb/cockroach/blob/0017b1c26904f8168b7581528f59264edfa9e881/pkg/sql/sem/builtins/generator_builtins.go#L260)):*
> ```Go
> // with integer bounds.
> type seriesValueGenerator struct {
> 	ctx                                 *tree.EvalContext
> ```

It'd be better and less error-prone to change the interface of Next() to take an evalcontext as argument, than to store a reference to it here.


<!-- Sent from Reviewable.io -->

bobvawter reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31146)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/settings/cluster/cockroach_versions.go, line 288 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/31146#-LOQJHfS9MbUUBOz0398-r2-288:-LOUTmYEA-6ulS8Ty3R-:b-9qyyv5) ([raw file](https://github.com/cockroachdb/cockroach/blob/0017b1c26904f8168b7581528f59264edfa9e881/pkg/settings/cluster/cockroach_versions.go#L288)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I think a distsql version is a good idea, which will prevent inconsistency within a session. 
</blockquote></details>

Ah, I see now that's actually a separate thing from the cluster version.  Will revise.


<!-- Sent from Reviewable.io -->

bobvawter committed:
sql: Align timestamp math with PostgreSQL

Our implementation of time+duration math delegates to golang's
time.Time.AddDate(), which has different normalization semantics when compared
to PostgreSQL when normalizing invalid month-of-day values.

The most likely input case is that the duration will have no month component
and we can fall back to go's implementation.  Another fast-path of checking the
day-of-month allows us to skip the special-case logic for start dates that
aren't the 29th, 30th, or 31st of a month.

```
name                                      old time/op    new time/op    delta
Add/fast-path-by-no-months-in-duration-8    63.2ns ± 4%    64.8ns ± 1%      ~     (p=0.151 n=5+5)
Add/fast-path-by-day-number-8               70.5ns ± 5%    87.5ns ± 0%   +24.12%  (p=0.008 n=5+5)
Add/no-adjustment-8                         70.6ns ± 3%   196.4ns ± 0%  +178.11%  (p=0.008 n=5+5)
Add/with-adjustment-8                       68.1ns ± 3%   218.2ns ± 1%  +220.60%  (p=0.008 n=5+5)

(No difference in allocations)
```

Resolves #30976

Release note (backward-incompatible change): The normalized results of
certain timestamp + duration operations involving year or month durations have
been adjusted to agree with the values returned by PostgreSQL.

Release note (sql change): In cases such as '2018-01-31'::TIMESTAMP + '1
month', where an intermediate result of February 31st needs to be normalized,
previous versions of CockroachDB would advance to March 3.  Instead, we would
now "round down" to February 28th to agree with the values returned by
PostgreSQL.  This change also affects the results of `generate_sequence()` when
used with timestamps.

bobvawter reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31146)* status: :shipit: complete! 0 of 0 LGTMs obtained

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/31146#-LOYT1vkBqbkP-cAF7-0:-LOYT1vkBqbkP-cAF7-1:b-3a351j) (no related file):*
PTAL: Now with a new distsql version check to enable legacy mode.

---
___
*[pkg/sql/sem/builtins/generator_builtins.go, line 260 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/31146#-LOUSzKB8MWLq2r_PSxU:-LOUakAPAJZDWjGn-PH3:b-h6194y) ([raw file](https://github.com/cockroachdb/cockroach/blob/0017b1c26904f8168b7581528f59264edfa9e881/pkg/sql/sem/builtins/generator_builtins.go#L260)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

It'd be better and less error-prone to change the interface of Next() to take an evalcontext as argument, than to store a reference to it here.
</blockquote></details>

That was my first thought, but it turns into a serious sprawl through a whole lot of other value-generation methods, since we would wind up having to also change the `sql.valueIterator` and `sql.planNode` interfaces in order to update `projectSetNode.Next()`.


<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM (from a versioning perspective; I didn't look at the implementation closely)

bobvawter commented:
bors r+

bobvawter commented:
Retrying bors due to setup failure in orm tests; looks like pip couldn't pull down a package?

bors r+

bobvawter commented:
orm passed in retry, now roachtest flaked (was ok in first build).  Retrying.

bors r+

knz commented:
today is not your lucky day it seems...
perhaps try a rebase?

bobvawter commented:
Retrying after rebase

bors r+
